File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    answer = len(s)
    if len(s) <= 2:
        return len(s)

    for i in range(1, len(s) // 2 + 1):
        parts = []
        idx = 0
        while idx + i <= len(s):
            if parts and parts[-1][0] == s[idx:idx + i]:
                parts[-1][1] += 1
            else:
                parts.append([s[idx:idx + i], 1])
            idx += i

        compressed_len = len(s[idx:])
        for part in parts:
            compressed_len += i
            if part[1] > 1:
                compressed_len += len(str(part[1]))
        answer = min(answer, compressed_len)
        
    return answer

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.util.ArrayList;

class Solution {
	public int solution(String s) {
		int answer = s.length();
		int len = s.length();

		for (int i = 1; i <= (int) len / 2; i++) {
			ArrayList<String> split = new ArrayList<>();

			int idx = 0;
			while (idx < len) {
				split.add(s.substring(idx, Math.min(len, idx + i)));
				idx += i;
			}

			ArrayList<Integer> compressedLen = new ArrayList<>();
			int compressedCnt = 0;
			for (int j = 0; j < split.size(); j++) {
				if (j != 0) {
					if (split.get(j - 1).equals(split.get(j))) {
						compressedCnt++;
					} else {
						if (compressedCnt > 0) {
							compressedLen.add(++compressedCnt);
							compressedCnt = 0;
						}

					}
				}
			}
			if (compressedCnt > 0) {
				compressedLen.add(++compressedCnt);
				compressedCnt = 0;
			}

			int compressedLength = s.length();

			for (int comp : compressedLen) {
				compressedLength -= (i * comp);
				compressedLength += (Integer.toString(comp).length() + i);
			}

			answer = Math.min(answer, compressedLength);

		}

		return answer;
	}
}